File: 2020/day03.py
#%%
import numpy as np

from utils import load

raw = load("day03.txt")
block = len(raw[0])

# %%
def get_tree(x, y):
    return raw[y][x % block]


def run(modx, mody):
    count = 0
    for i in range(len(raw)):
        y = mody * i
        if y < len(raw):
            x = modx * i
            if get_tree(x, y) == "#":
                count += 1
    return count


# %%
def test1():
    assert run(3, 1) == 265


def test2():
    assert np.prod([run(i, 1) for i in range(1, 8, 2)]) * run(1, 2) == 3154761400


# %%

File: 2020/rust/src/day03.rs
use super::utils::*;

fn run(raw: &Vec<String>, modx: usize, mody: usize) -> usize {
    let block_size = raw[0].len();
    let mut count: usize = 0;
    for i in 0..raw.len() {
        let y = mody * i;
        if y < raw.len() && str_idx(&raw[y], (modx * i) % block_size) == '#' {
            count += 1;
        }
    }
    count
}

pub fn part1(input: &Vec<String>) -> usize {
    run(input, 3, 1)
}

pub fn part2(input: &Vec<String>) -> usize {
    run(input, 1, 2) * (1..8).step_by(2).map(|i| run(&input, i, 1)).product::<usize>()
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]

    fn test1() {
        assert_eq!(part1(&read("day03.txt")), 265);
    }

    #[test]
    fn test2() {
        assert_eq!(part2(&read("day03.txt")), 3154761400);
    }
}
